File: Projects/Walmart/eval.py
# -*- coding:utf-8 -*-
# =========================================================================== #
# Project : Walmart Sales Prediction Model                                    #
# File    : \eval.py                                                          #
# Python  : 3.9.1                                                             #
# --------------------------------------------------------------------------- #
# Author  : John James                                                        #
# Company : nov8.ai                                                           #
# Email   : john.james@nov8.ai                                                #
# URL     : https://github.com/john-james-sf/Practical-Statistical-Learning   #
#           /Projects/Walmart/                                                #
# --------------------------------------------------------------------------- #
# Created       : Saturday, April 3rd 2021, 5:41:24 pm                        #
# Last Modified : Saturday, April 3rd 2021, 5:41:24 pm                        #
# Modified By   : John James (john.james@nov8.ai)                             #
# --------------------------------------------------------------------------- #
# License : BSD                                                               #
# Copyright (c) 2021 nov8.ai                                                  #
# =========================================================================== #
import numpy as np
import pandas as pd

from mymain import mypredict

train = pd.read_csv('train_ini.csv', parse_dates=['Date'])
test = pd.read_csv('test.csv', parse_dates=['Date'])

# save weighed mean absolute error WMAE
n_folds = 10
next_fold = None
wae = []

# time-series CV
for t in range(1, n_folds+1):
    print(f'Fold{t}...')

    # *** THIS IS YOUR PREDICTION FUNCTION ***
    train, test_pred = mypredict(train, test, next_fold, t)

    # Load fold file
    # You should add this to your training data in the next call to mypredict()
    fold_file = 'fold_{t}.csv'.format(t=t)
    next_fold = pd.read_csv(fold_file, parse_dates=['Date'])

    # extract predictions matching up to the current fold
    scoring_df = next_fold.merge(test_pred, on=['Date', 'Store', 'Dept'], how='left')

    # extract weights and convert to numpy arrays for wae calculation
    weights = scoring_df['IsHoliday_x'].apply(lambda is_holiday:5 if is_holiday else 1).to_numpy()
    actuals = scoring_df['Weekly_Sales'].to_numpy()
    preds = scoring_df['Weekly_Pred'].fillna(0).to_numpy()

    wae.append((np.sum(weights * np.abs(actuals - preds)) / np.sum(weights)).item())

print(wae)
print(sum(wae)/len(wae))

File: Projects/Walmart/eval.R
source("mymain.R")
library(lubridate)

# Parameters for models.
m0 <- list(model="regression", data='r', adjust=FALSE, model.type=NULL, n.comp=NULL)
m1 <- list(model="regression", data='r', adjust=TRUE, model.type=NULL, n.comp=NULL)
m2 <- list(model="snaive", data='t', adjust=TRUE, model.type=NULL, n.comp=NULL)
m3 <- list(model="snaive.custom", data='t', adjust=TRUE, model.type=NULL, n.comp=NULL)
m4 <- list(model="stlf", data='t', adjust=TRUE, model.type="ets", n.comp=NULL)
m5 <- list(model="stlf", data='t', adjust=TRUE, model.type="arima", n.comp=NULL)
m6 <- list(model="stlf.svd", data='t', adjust=TRUE, model.type="ets", n.comp=12)
m7 <- list(model="stlf.svd", data='t', adjust=TRUE, model.type="arima", n.comp=12)
m8 <- list(model="tslm", data='t', adjust=TRUE, model.type=NULL, n.comp=NULL)
m9 <- list(model="tslm.svd", data='t', adjust=TRUE, model.type="ets", n.comp=12)
m10 <- list(model="tslm.svd", data='t', adjust=TRUE, model.type="arima", n.comp=12)

params=NULL
params <- m0
# read in train / test dataframes
train <- readr::read_csv('data/train_ini.csv')
test <- readr::read_csv('data/test.csv')

# save weighted mean absolute error WMAE
num_folds <- 10
wae <- rep(0, num_folds)

# Create lists for metadata
folds <- list()
scores <- list()
times <- list()

model_start <- Sys.time()

for (t in 1:num_folds) {
  fold_start <- Sys.time()    
  # *** THIS IS YOUR PREDICTION FUNCTION ***  
  test_pred <- mypredict(params)
  
  # load fold file 
  fold_file <- paste0('data/fold_', t, '.csv')
  new_train <- readr::read_csv(fold_file, 
                              col_types = cols())

  # extract predictions matching up to the current fold
  scoring_tbl <- new_train %>% 
      left_join(test_pred, by = c('Date', 'Store', 'Dept'))
  
  # compute WMAE
  actuals <- scoring_tbl$Weekly_Sales
  preds <- scoring_tbl$Weekly_Pred
  preds[is.na(preds)] <- 0
  weights <- if_else(scoring_tbl$IsHoliday, 5, 1)
  wae[t] <- sum(weights * abs(actuals - preds)) / sum(weights)

  # Store process metadata for reporting
  fold_end <- Sys.time()
  fold_time <- fold_end - fold_start
  fold_time <- round(fold_time, digits=2)
  # Create metadata
  folds[[t]] <- t
  scores[[t]] <- wae[t]
  times[[t]] <- fold_time    

  x <- sprintf("Model %s: Fold: %d  WAE: %f  Elapsed: %f", toupper(params$model), t, wae[t], fold_time)
  print(x)
}
# Save Model Statistics
stats <- list("Fold"=folds, "WAE"=scores, "Time Elapsed"=times)
save_stats(stats, params)
# Report model summary
model_end <- Sys.time()
model_time <- model_end - model_start
model_time <- round(model_time,digits=2)
summary.stats <- list(params=params, score=mean(wae), time=model_time)
save_summary_stats(summary.stats, params)
x <- sprintf("Model %s: Mean WAE: %f  Elapsed: %f", toupper(params$model), mean(wae), model_time)
print(x)


